File: 02.Fila_Prioridade/Quicksort.java
import java.util.Random;

public class Quicksort<T extends Comparable<T>> {
    private T data[];

    public void sort(T[] data) {
        this.data = data;
        sort(0, data.length-1);
    }

    public void sort(int p, int r) {
        if(p < r) {
            int q = partition(p, r);
            sort(p, q-1);
            sort(q+1, r);
        }
    }

    private int partition(int p, int r) {
        int q = p;
        for(int j=p; j<r; j++) {
            if(data[j].compareTo(data[r]) < 0) {
                swap(j,q);
                q++;
            }
        }
        swap(r,q);
        return q;
    }

    private void swap(int p1, int p2) {
        T tmp = data[p1];
        data[p1] = data[p2];
        data[p2] = tmp;
    }

    public static void main(String args[]) {
        Random r = new Random();
        int MAX = 100000;
        for(int n=500; n<=MAX; n+=500) {
            Integer[] data = new Integer[n];
            for(int i=0; i<data.length; i++)
                data[i] = r.nextInt(data.length*10);

            Quicksort<Integer> qs = new Quicksort<>();
            long start = System.nanoTime();
            qs.sort(data);
            long end = System.nanoTime();
            double tempo = (end-start)/1e9;
            System.out.printf("%d %f\n",n,tempo);
            //for(int i=0; i<data.length; i++)
            //    System.out.print(data[i]+" ");
            //System.out.println();
        }
    }
}

File: 02.Fila_Prioridade/quicksort.py
#!/usr/bin/python3
#
# Quicksort
#

import random
from time import process_time

class Quicksort:

    def partition(self, data, p, r):

        q = p
        for j in range(p,r):
            if data[j] <= data[r]:
                data[j], data[q] = data[q], data[j]
                q += 1
        data[r], data[q] = data[q], data[r]
        return q


    def sort(self, data):
        self.__quicksort(data, 0, len(data)-1)

    def __quicksort(self, data, p, r):
        if p < r:
            q = self.partition(data, p, r)
            self.__quicksort(data, p, q-1)
            self.__quicksort(data, q+1, r)


if __name__ == "__main__":
    MAX = 100000
    for n in range(500,MAX,500):
        data = [random.randint(0,MAX*10) for x in range(n)]

        qs = Quicksort()
        start = process_time()
        qs.sort(data)
        end = process_time()
        t = end-start
        print(f"{n} {t}")

    #print(f"Tempo para ordenar {MAX} elementos: {t} s")

